cpp/dolfinx/mesh/graphbuild.cpp

```diff
@@ -201,103 +198,86 @@ compute_nonlocal_dual_graph(
+  for (std::size_t i = 0; i < returned_list.size(); ++i)
   {
-    if (facet_match[i])
+    const std::int64_t cell1 = returned_list[i];
+    if (cell1 != -1)
     {
-      const int i0 = perm[i - 1];
-      const int i1 = perm[i];
-      const int proc0 = proc[i0];
-      const int proc1 = proc[i1];
-      const auto facet0 = recvd_buffer.links(i0);
-      const auto facet1 = recvd_buffer.links(i1);
-
-      const std::int64_t cell0 = facet0.back();
-      const std::int64_t cell1 = facet1.back();
-
-      auto buffer0 = send_buffer.links(proc0);
-      buffer0[pos[proc0]++] = cell0;
-      buffer0[pos[proc0]++] = cell1;
-      auto buffer1 = send_buffer.links(proc1);
-      buffer1[pos[proc1]++] = cell1;
-      buffer1[pos[proc1]++] = cell0;
+      const std::int64_t cell0 = send_list[i * (max_num_vertices_per_facet + 1)
+                                           + max_num_vertices_per_facet]
+                                 - cell_offset;
+      boundary_vertices.insert(
+          boundary_vertices.end(),
+          std::next(send_list.begin(), i * (max_num_vertices_per_facet + 1)),
+          std::next(send_list.begin(), i * (max_num_vertices_per_facet + 1)
+                                           + max_num_vertices_per_facet));
+      ++edge_count[cell0];
+      ++num_ghost_edges;
     }
   }
-
-  // Send matches to other processes
-  const std::vector<std::int64_t> cell_list
-      = dolfinx::MPI::all_to_all(comm, send_buffer).array();
-
-  // Ghost nodes: insert connected cells into local map
-
-  // Count number of adjacency list edges
-  std::vector<int> edge_count(local_graph.num_nodes(), 0);
-  for (int i = 0; i < local_graph.num_nodes(); ++i)
-    edge_count[i] += local_graph.num_links(i);
-  for (std::size_t i = 0; i < cell_list.size(); i += 2)
-  {
-    assert(cell_list[i] - cell_offset >= 0);
-    assert(cell_list[i] - cell_offset < std::int64_t(edge_count.size()));
-    edge_count[cell_list[i] - cell_offset] += 1;
-  }
+  // Get unique set of vertices
+  std::sort(boundary_vertices.begin(), boundary_vertices.end());
+  boundary_vertices.erase(
+      std::unique(boundary_vertices.begin(), boundary_vertices.end()),
+      boundary_vertices.end());
 
   // Build adjacency list
   offsets.assign(edge_count.size() + 1, 0);
   std::partial_sum(edge_count.begin(), edge_count.end(),
                    std::next(offsets.begin()));
   graph::AdjacencyList<std::int64_t> graph(
       std::vector<std::int64_t>(offsets.back()), std::move(offsets));
+
+  // Copy existing local graph, adding cell_offset
   pos.assign(graph.num_nodes(), 0);
-  std::vector<std::int64_t> ghost_edges;
   for (int i = 0; i < local_graph.num_nodes(); ++i)
   {
-    auto local_graph_i = local_graph.links(i);
-    auto graph_i = graph.links(i);
-    for (std::size_t j = 0; j < local_graph_i.size(); ++j)
-      graph_i[pos[i]++] = local_graph_i[j] + cell_offset;
+    std::transform(local_graph.links(i).begin(), local_graph.links(i).end(),
```

Comment:
Create `local_graph.links(i)` before std::transform to avoid multiple constructions?

(might be handy to add some iterators to Adjacency list)

Reply:
Not sure I understand what you mean. But  `pos` could be created using `std::adjacent_difference`, I think.

Reply:
`local_graph.links(i).begin()` and `local_graph.links(i).end()` creates span twice.